### PR TITLE
Themes: Decouple client `router` and `renderer`

### PR DESCRIPTION
**Pull request by ockham**
Continuing the discussion from https://github.com/Automattic/wp-calypso/pull/3397#issuecomment-186373943 with this PR as a suggestion.

The main rationale here is separation of concerns, visibility, and DevX for folks writing section specific middleware. Before, `renderer` was just magically appended to the end of each route, adding another constraint to the way section middleware had to be written (i.e. less separation of concern -- e.g. catch-all routes -- was possible). Now, both `router` and `renderer` are building blocks that are available to section devs, and it's up to them how to use them. (This way, they don't actually need to render in every case -- they can conceivably even `redirect`. Also, different `render` functions can be passed to sections depending on whether we're logged-in or out.)

To test -- verify that everything still works as before.

/cc @mcsf @ehg @seear

### CODE REVIEW

**Comment by seear** [comment]
> To test -- verify that everything still works as before.

Is it possible to narrow the scope here a bit? Is this only /design logged-out?

**Comment by ockham** [comment]
> Is it possible to narrow the scope here a bit?

I'm afraid no, since all our `/design` and `/themes` routes are affected, both logged-out and -in.

File: server/bundler/loader.js [hunk 1]
```diff
@@ -86,7 +86,7 @@ function splitTemplate( path, module, chunkName ) {
 		'		}',
 		'		context.store.dispatch( { type: "SET_SECTION", isLoading: false } );',
 		'		if ( ! _loadedSections[ ' + JSON.stringify( module ) + ' ] ) {',
-		'			require( ' + JSON.stringify( module ) + ' )( controller.clientRouter );',
+		'			require( ' + JSON.stringify( module ) + ' )( controller.clientRouter, controller.clientRenderer );',
```
Comment (seear) [inline server/bundler/loader.js]: Naive question: what happens here for controllers that don't have these functions?

Comment (ockham) [inline server/bundler/loader.js]: Not sure I understand the question... We're passing these functions here as arguments for indiviual sections' controllers to use, but they can entirely ignore them (all sections except for ours do so far). [`controller`](https://github.com/Automattic/wp-calypso/blob/update/decouple-client-router-and-renderer/server/bundler/loader.js#L16) here, OTOH, is `client/controller.js`, which exports `clientRouter` and `clientRenderer`.

Comment (seear) [inline server/bundler/loader.js]: > is client/controller.js, which exports clientRouter and clientRenderer

Got it, cheers :+1:

File: client/controller.js [hunk 2]
```diff
@@ -26,10 +26,10 @@ const debug = debugFactory( 'calypso:controller' );
  * divs.
  */
 export function clientRouter( route, ...middlewares ) {
-	page( route, ...[ ...middlewares, render ] );
+	page( route, ...middlewares );
```
Comment (ockham) [inline client/controller.js]: Heh, I should just `export const clientRouter = page`.

File: client/my-sites/themes/index.js [hunk 3]
```diff
@@ -6,37 +6,29 @@ import userFactory from 'lib/user';;
...
+		if ( isLoggedIn ) {
+			router( '/design', multiSite, navigation, siteSelection );
+			router( '/design/:site_id', singleSite, navigation, siteSelection );
+			router( '/design/type/:tier', multiSite, navigation, siteSelection );
+			router( '/design/type/:tier/:site_id', singleSite, navigation, siteSelection );
+		} else {
+			router( '/design', loggedOut, makeLoggedOutLayout );
+			router( '/design/type/:tier', loggedOut, makeLoggedOutLayout );
+		}
+		router( '/design*', renderer );
```
Comment (seear) [inline client/my-sites/themes/index.js]: Is there a scenario where one would _not_ hook up the passed in renderer?

**Comment by ockham** [comment]
Pushed two more commits to fix JSDoc and export `page` in a simpler way.

File: client/my-sites/themes/index.js (hunk 3 above)
Comment (ockham) [inline client/my-sites/themes/index.js]: Redirects, possibly?

What's more important is that this allows better separation of concerns with routing. Up to this PR, the renderer was magically attached as last middleware in each individual `router` invocation. That's an extra constraint imposed on individual sections' developers to keep in mind.

Now, we should be able to do things like

```js
if ( isLoggedIn ) {
	router( '/design', multiSite );
	router( '/design/:site_id', singleSite );
	router( '/design/type/:tier', multiSite );
	router( '/design/type/:tier/:site_id', singleSite );
	router( '/design*', navigation, siteSelection );
}
```

again, i.e. use catch-all middleware that is invoked for all subroutes -- and other sections' authors possibly relying on that `page` feature don't have to change their routing round.

(Maybe I should actually push a commit perusing that catch-all technique again?)

**Comment by seear** [comment]
The contract for implementing `function( router, renderer ) ` is:
If you use the router that is passed in, you have to make sure that you call it, finally, for every concerned route, with either the passed in renderer or something else.

Is that made explicit somewhere, or can it only be inferred by reading the existing example in `/themes/index.js`?

**Comment by ockham** [comment]
Not very explicit yet, and it might make sense to add a doc on isorouting somewhere.

(However, it's also implicitly understood now that sections use `ReactDom.render()` to e.g. render to `primary` as they see fit, without that being documented much AFAICT. And they could actually continue to do so -- they just wouldn't be able to leverage isorouting then.

IMO the beauty of the middleware approach is that you can hook in at whatever level you prefer -- you can pick and choose from using `makeLayout` to get layout-last, `router` and `renderer` to get isorouting, but you can also just go ahead and continue to render like you did before.)

**Comment by ockham** [comment]
One more argument that just occurred to me -- in terms of a mental model, it's probably easier to think of `router` as 'works-like-`page`' than 'works-like-`page`-but-with-rendering-middleware-attached'...

**Comment by apeatling** [comment]
@ockham @seear Is this something that will get merged? If not it's been a long time and we should close this.

**Comment by ockham** [comment]
Okay, closing for now.

**Pull request closed without merging**